src/cart/crt_iv.c

```diff
@@ -2757,7 +2758,8 @@ handle_response_cb(const struct crt_cb_info *cb_info)
 		return;
 	}
 callback:
-	handle_response_internal((void *)cb_info);
+	handle_response_internal(cb_info2);
+	D_FREE(cb_info2);
```

Comment:
We should not be freeing cb_info2 ourselves here

Reply:
Hmm. Where it's free? I was not able to find.

Reply:
cb_info is passed by ref by the cart internal code from the callers stack; it does not get allocated. For example in crt_req_send():
1326                         struct crt_cb_info      cbinfo;
1327
1328                         cbinfo.cci_rpc = NULL;
1329                         cbinfo.cci_arg = arg;
1330                         cbinfo.cci_rc  = -DER_INVAL;
1331                         complete_cb(&cbinfo);

Reply:
In good path it will be free in handle_response_cb_internal() but in case of error to call this callback it will be free here.

Reply:
i think tihs requires a different fix then. this function is specified as direct handle completion in other cases for example:
1215         rc = crt_req_send(rpc, handle_response_cb, cb_info);

2336         rc = crt_req_send(corpc_req, handle_response_cb, iv_sync_cb);

2670         rc = crt_req_send(rpc, handle_response_cb, cb_info);

- in all these cases the cb_info will be passed as a pointer to structure on a stack when handle response cb executes

Reply:
the error is handled already on 2755, no?

Reply:
No, info != cb_info. info is allocated in L2737

Reply:
Correct, so you only need to cleanup that allocated info for the case of the deferred response callback processing. for the case of direct response callback processing the info is passed as a pointer to stack structure

Reply:
From my observation all cb_info structures which are passed to this function are dynamically allocated.

Reply:
Lets try to find all places when handle_response_cb is specified with appropriate cb_info.
There are three places L1215, L2336 and L2670. Then try to find which cb_info is used for those calls.

For L1215 its go from parameters of crt_ivf_rpc_issue() which is called from L1445 and L1749. For first L1445 call cb_info allocated in L1426. For second L1749 call it's allocated in L1730.

For L2336 cb_info is allocated in L2285.

For L2670 its also go from parameters of crt_ivu_rpc_issue() which is called from L2849, L3109 and L3355. For first L2849 call cb_info allocated in L2814. For second L3109 call it's allocated in L3092. For third L3355 call it's allocated in L3340.

As you can see in all places it's dynamically allocated. So, it should be free. Could you specify the path when this struction is in stack?

Reply:
Lets start with L1251:
1215 rc = crt_req_send(rpc, handle_response_cb, cb_info);

note that this cb_info that what is called in handle_response_cb(const struct crt_cb_info *cb_info)  are two different structures.
crt_rpc.c:
1318 crt_req_send(crt_rpc_t *req, crt_cb_t complete_cb, void *arg)
...
1326                         struct crt_cb_info      cbinfo;
1327
1328                         cbinfo.cci_rpc = NULL;
1329                         cbinfo.cci_arg = arg;
1330                         cbinfo.cci_rc  = -DER_INVAL;
1331                         complete_cb(&cbinfo);

cb_info passed on line 1215 of crt_iv.c is cb_info->cci_arg inside of handle_response_cb()

This cb_info->cci_arg will be freed in (since this is fetch operation):
handle_response_internal->handle_ivfetch_response->
1043         struct iv_fetch_cb_info         *iv_info = cb_info->cci_arg;
...
1104         D_FREE(iv_info);
similar free-ing of cbi_info->cci_arg are done in other completion handlers:
2587         D_FREE(iv_info);
2193         D_FREE(iv_sync);

Reply:
Hmm. It looks you right.. This is very complicated. I was not able to catch this logic. Will remove it. Thanks!